subsys/net/lib/dhcpv4/dhcpv4.c

```diff
@@ -62,6 +64,45 @@ static inline bool dhcpv4_add_cookie(struct net_pkt *pkt)
 	return true;
 }
 
+#if defined(CONFIG_NET_DHCPV4_OPTION_CALLBACKS)
+static uint8_t *dhcpv4_option_callback_get_unique_types(uint8_t additional_types)
+{
+	struct net_dhcpv4_option_callback *cb, *tmp;
+	uint8_t *types;
+	int count = 0;
+	bool found = false;
+
+	size_t callback_syslist_len = sys_slist_len(&option_callbacks);
+
+	types = (uint8_t *)malloc((callback_syslist_len + additional_types) * sizeof(uint8_t));
```

Comment:
I'm not really in favor of introducing heap dependency to the DHCP client. Perhaps we could specify the maximum number of requested options at Kconfig level, and predefine an array on stack (or as a part of the dhcpv4 context struct)?